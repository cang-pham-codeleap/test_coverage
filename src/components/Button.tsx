/**
 * Button component with customizable variants
 * - Simple reusable button with primary/secondary variants
 * - Includes disabled state handling
 * - Uses type-safe props
 */

import React from "react";

type ButtonVariant = "primary" | "secondary";

interface ButtonProps {
  children: React.ReactNode;
  onClick?: () => void;
  disabled?: boolean;
  variant?: ButtonVariant;
  className?: string;
  "data-testid"?: string;
}

export function Button({
  children,
  onClick,
  disabled = false,
  variant = "primary",
  className = "",
  "data-testid": testId = "button",
}: ButtonProps) {
  // Base classes for all button variants
  const baseClasses =
    "px-4 py-2 rounded transition-colors focus:outline-none focus:ring-2";

  // Variant-specific classes
  const variantClasses = {
    primary: "bg-blue-500 text-white hover:bg-blue-600 focus:ring-blue-300",
    secondary:
      "bg-gray-200 text-gray-800 hover:bg-gray-300 focus:ring-gray-400",
    tertiary: "bg-gray-200 text-gray-800 hover:bg-gray-300 focus:ring-gray-400",
    quaternary:
      "bg-gray-200 text-gray-800 hover:bg-gray-300 focus:ring-gray-400",
    quinary: "bg-gray-200 text-gray-800 hover:bg-gray-300 focus:ring-gray-400",
    senary: "bg-gray-200 text-gray-800 hover:bg-gray-300 focus:ring-gray-400",
  };

  // Disabled state classes
  const disabledClasses = "opacity-50 cursor-not-allowed";

  // Combine all classes based on props
  const buttonClasses = `
    ${baseClasses} 
    ${variantClasses[variant]} 
    ${disabled ? disabledClasses : ""} 
    ${className}
  `.trim();

  return (
    <button
      onClick={onClick}
      disabled={disabled}
      className={buttonClasses}
      data-testid={testId}
    >
      {children}
    </button>
  );
}

export default Button;
