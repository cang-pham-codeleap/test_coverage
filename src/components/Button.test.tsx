/**
 * Button component tests
 * - Tests basic rendering
 * - Tests different variants
 * - Tests disabled state
 * - Tests click handling
 */

import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Button from "./Button";

describe("Button Component", () => {
  // Test basic rendering
  test("renders button with children", () => {
    render(<Button>Click me</Button>);
    const buttonElement = screen.getByTestId("button");
    expect(buttonElement).toMatchSnapshot();
  });

  // Test primary variant (default)
  test("renders primary variant by default", () => {
    render(<Button>Primary</Button>);
    const buttonElement = screen.getByTestId("button");
    // Check for primary class existence based on our implementation
    expect(buttonElement.className).toContain("bg-blue-500");
  });

  // Test secondary variant
  test("renders secondary variant when specified", () => {
    render(<Button variant="secondary">Secondary</Button>);
    const buttonElement = screen.getByTestId("button");
    // Check for secondary class existence
    expect(buttonElement.className).toContain("bg-gray-200");
  });

  // Test disabled state
  test("applies disabled attribute when disabled prop is true", () => {
    render(<Button disabled>Disabled</Button>);
    const buttonElement = screen.getByTestId("button");
    expect(buttonElement).toMatchSnapshot();
    expect(buttonElement.className).toContain("opacity-50");
  });

  // Test click handler
  test("calls onClick handler when clicked", () => {
    const handleClick = jest.fn();
    render(<Button onClick={handleClick}>Click me</Button>);
    const buttonElement = screen.getByTestId("button");

    fireEvent.click(buttonElement);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  // Test click handler not called when disabled
  test("does not call onClick when disabled", () => {
    const handleClick = jest.fn();
    render(
      <Button onClick={handleClick} disabled>
        Click me
      </Button>
    );
    const buttonElement = screen.getByTestId("button");

    fireEvent.click(buttonElement);
    expect(handleClick).not.toHaveBeenCalled();
  });

  // Test custom class name
  test("applies custom className when provided", () => {
    const customClass = "my-custom-class";
    render(<Button className={customClass}>Custom Style</Button>);
    const buttonElement = screen.getByTestId("button");

    expect(buttonElement.className).toContain(customClass);
  });

  // Test custom test ID
  test("applies custom data-testid when provided", () => {
    render(<Button data-testid="custom-button">Custom TestID</Button>);
    const buttonElement = screen.getByTestId("custom-button");

    expect(buttonElement).toMatchSnapshot();
  });
});
