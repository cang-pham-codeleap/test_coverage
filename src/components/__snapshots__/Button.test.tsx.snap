// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component applies custom data-testid when provided 1`] = `
<button
  class="px-4 py-2 rounded transition-colors focus:outline-none focus:ring-2 
    bg-blue-500 text-white hover:bg-blue-600 focus:ring-blue-300"
  data-testid="custom-button"
>
  Custom TestID
</button>
`;

exports[`Button Component applies disabled attribute when disabled prop is true 1`] = `
<button
  class="px-4 py-2 rounded transition-colors focus:outline-none focus:ring-2 
    bg-blue-500 text-white hover:bg-blue-600 focus:ring-blue-300 
    opacity-50 cursor-not-allowed"
  data-testid="button"
  disabled=""
>
  Disabled
</button>
`;

exports[`Button Component renders button with children 1`] = `
<button
  class="px-4 py-2 rounded transition-colors focus:outline-none focus:ring-2 
    bg-blue-500 text-white hover:bg-blue-600 focus:ring-blue-300"
  data-testid="button"
>
  Click me
</button>
`;
